File: frontend/src/utils/recommendations.js
export const ACTION_COPY = {
  allow: {
    label: 'Allow',
    short: 'needed',
    tone: 'text-success',
    border: 'var(--success)',
    description: 'Keep visible because it is required for this purpose.',
  },
  mask: {
    label: 'Mask',
    short: 'partial',
    tone: 'text-warning',
    border: 'var(--warning)',
    description: 'Hide most of the value while leaving limited proof visible.',
  },
  redact: {
    label: 'Redact',
    short: 'remove',
    tone: 'text-danger',
    border: 'var(--danger)',
    description: 'Remove from the shared copy because it is excess or high-risk.',
  },
}

export const CONTEXT_GUIDANCE = {
  age_verification: {
    title: 'Age verification needs minimum proof only',
    body: 'Recommended defaults usually mask DOB, allow name if required, and redact ID numbers, address, QR, photo, and signature.',
  },
  identity_verification: {
    title: 'Identity verification should not expose the full document',
    body: 'Recommended defaults usually allow name and DOB, mask document numbers, and redact address, QR, MRZ, photo, signature, and unrelated fields.',
  },
  address_proof: {
    title: 'Address proof should focus on residence',
    body: 'Recommended defaults usually allow name and address, and redact document numbers, DOB, gender, photo, QR, MRZ, and signature.',
  },
  kyc_onboarding: {
    title: 'KYC needs more fields but still benefits from masking',
    body: 'Recommended defaults usually allow required identity/address fields, mask document numbers, and redact photo, QR, MRZ, signature, and extra fields.',
  },
  general_upload: {
    title: 'General upload has no declared necessity',
    body: 'Recommended defaults redact detected PII because no specific purpose justifies sharing personal fields.',
  },
}

export function recommendationCounts(fields = []) {
  return fields.reduce((acc, field) => {
    const action = field.suggested_action || field.redaction_decision || 'redact'
    acc[action] = (acc[action] || 0) + 1
    return acc
  }, { allow: 0, mask: 0, redact: 0 })
}

export function actionCopy(action) {
  return ACTION_COPY[action] || ACTION_COPY.redact
}

File: securemask/core/recommendations.py
"""Context-aware redaction recommendations."""
from __future__ import annotations

from dataclasses import dataclass

from securemask.models.detected_field import DetectedField


from securemask.config import IDENTIFIER_FIELDS

VISUAL_SECRET_FIELDS = {"photo", "signature", "qr_code", "mrz_lines"}
AGE_FIELDS = {"dob", "date_of_birth"}


@dataclass(frozen=True)
class Recommendation:
    action: str
    reason: str


def recommend_action(field: DetectedField, document_type: str, context: str, required: bool) -> Recommendation:
    """Return the safest default user action for a field.

    Actions intentionally stay within the existing UI contract:
    ``allow``, ``mask``, or ``redact``.
    """
    name = field.field_name

    if field.always_redact or name in VISUAL_SECRET_FIELDS:
        return Recommendation(
            "redact",
            "Remove this visual or machine-readable region because it can expose the full identity even when text fields are hidden.",
        )

    if name in IDENTIFIER_FIELDS:
        if required and context in {"identity_verification", "kyc_onboarding"}:
            return Recommendation(
                "mask",
                "Keep only enough of the identifier visible for verification and hide the rest to reduce replay or fraud risk.",
            )
        return Recommendation(
            "redact",
            "This identifier is not necessary for the selected purpose, so the safest option is to remove it.",
        )

    if name in AGE_FIELDS and context == "age_verification":
        return Recommendation(
            "mask",
            "Age checks usually need proof of age, not the complete date of birth, so partial masking is safer.",
        )

    if required:
        return Recommendation(
            "allow",
            f"This field is needed for {context.replace('_', ' ')} on a {document_type.replace('_', ' ')} document.",
        )

    return Recommendation(
        "redact",
        f"This field is excess for {context.replace('_', ' ')}, so redaction follows data minimisation.",
    )


def summarize_recommendations(fields: list[DetectedField]) -> dict[str, int | str]:
    counts = {"allow": 0, "mask": 0, "redact": 0}
    for field in fields:
        action = field.suggested_action if field.suggested_action in counts else "redact"
        counts[action] += 1

    return {
        **counts,
        "summary": (
            f"Recommended defaults: allow {counts['allow']} required fields, "
            f"mask {counts['mask']} verification fields, redact {counts['redact']} excess or high-risk fields."
        ),
    }
